### PR TITLE
Consent journey email validation

### PR DESCRIPTION
**Pull request by Calum-Campbell**
### Two changes
1.
 - Link sent in the email validation should redirect to consents journey page
- The link will always go to the default which is just /consents
- Behind a switch as shouldn't be live in production until we want to expose the journey page. 
2.
- Now use endpoint syntax rather than query param syntax for consents journey (consents/all, /consents/newsletters, /consents)

Related PR: 
https://github.com/guardian/identity/pull/740
## What is the value of this and can you measure success?
- Repermission
## Does this affect other platforms - Amp, Apps, etc?
- Emails will be sent when registering in App

## Tested in CODE?
No

### CODE REVIEW

**Review by jfsoul — APPROVED** [review]

File: identity/app/controllers/EmailVerificationController.scala [hunk 1]
```diff
@@ -42,8 +46,13 @@ class EmailVerificationController(api: IdApiClient,
           val userIsLoggedIn = authenticationService.requestPresentsAuthenticationCredentials(request)
           val verifiedReturnUrlAsOpt = returnUrlVerifier.getVerifiedReturnUrl(request)
           val verifiedReturnUrl = verifiedReturnUrlAsOpt.getOrElse(returnUrlVerifier.defaultReturnUrl)
+          val encodedReturnUrl = URLEncoder.encode(verifiedReturnUrl, "utf-8")
 
-          Ok(views.html.emailVerified(validationState, page, idRequest, idUrlBuilder, userIsLoggedIn, verifiedReturnUrl))
+          if(validationState.isExpired || IdentityPointToConsentJourneyPage.isSwitchedOff) {
+            Ok(views.html.emailVerified(validationState, page, idRequest, idUrlBuilder, userIsLoggedIn, verifiedReturnUrl))
+          } else {
+            SeeOther(idUrlBuilder.buildUrl(s"/consents/?returnUrl=${encodedReturnUrl}"))
```
Comment (jfsoul) [inline identity/app/controllers/EmailVerificationController.scala]: I generally don't like to have a trailing slash before the `?` (or at all)

File: identity/conf/routes [hunk 2]
```diff
@@ -62,6 +62,7 @@ POST        /privacy/edit-ajax                      controllers.EditProfileContr
 GET         /email-prefs                            controllers.EditProfileController.displayEmailPrefsForm(consentsUpdated: Boolean ?= false, consentHint: Option[String])
 POST        /email-prefs                            controllers.EditProfileController.saveEmailPreferencesAjax
 
-GET         /consent                                controllers.EditProfileController.displayConsentJourneyForm(journey: Option[String], consentHint: Option[String], newsletterHint: Option[String])
-POST        /complete-consent                         controllers.EditProfileController.submitRepermissionedFlag
+GET         /consents                               controllers.EditProfileController.displayConsentJourneyForm(journey = "consent", consentHint: Option[String], newsletterHint: Option[String])
```
Comment (jfsoul) [inline identity/conf/routes]: it feels like these 2 routes are the same one (with optional journey and trailing slash) but not a big deal if fiddly to get working

File: identity/conf/routes [hunk 3]
```diff
@@ -62,6 +62,7 @@ POST        /privacy/edit-ajax                      controllers.EditProfileContr
...
+GET         /consents/:journey                      controllers.EditProfileController.displayConsentJourneyForm(journey: String, consentHint: Option[String], newsletterHint: Option[String])
+POST        /complete-consent                       controllers.EditProfileController.submitRepermissionedFlag
```
Comment (jfsoul) [inline identity/conf/routes]: do we need to pluralise this too? `complete-consents`

File: identity/test/controllers/EmailVerificationControllerTest.scala [hunk 4]
```diff
@@ -51,7 +51,7 @@ class EmailVerificationControllerTest extends path.FreeSpec
 
     "when the api call succeeds" - {
       when(api.validateEmail(token, trackingData)).thenReturn(Future.successful(Right(())))
-      val result = controller.verify(token)(testRequest)
+      val result = controller.verify(token, "EmailValidation")(testRequest)
```
Comment (jfsoul) [inline identity/test/controllers/EmailVerificationControllerTest.scala]: Are these changes needed any more?

File: identity/conf/routes (hunk 2 above)
Comment (Calum-Campbell) [reply identity/conf/routes]: Yep it's true it seems like duplication and doesn't look very neat, but from what I can tell it's the only way to have the routes like this, as you can pass query params as optional but not actual /routes/.

File: identity/test/controllers/EmailVerificationControllerTest.scala (hunk 4 above)
Comment (Calum-Campbell) [reply identity/test/controllers/EmailVerificationControllerTest.scala]: No, cheers.

File: identity/conf/routes (hunk 3 above)
Comment (Calum-Campbell) [reply identity/conf/routes]: I think we should yep

File: identity/conf/routes (hunk 2 above)
Comment (jfsoul) [reply identity/conf/routes]: You could do it with regex, but not important

File: identity/app/controllers/EditProfileController.scala [hunk 5]
```diff
@@ -29,7 +29,13 @@ object EmailPrefsProfilePage extends IdentityPage("/email-prefs", "Emails")
 object MembershipEditProfilePage extends IdentityPage("/membership/edit", "Membership")
 object recurringContributionPage extends IdentityPage("/contribution/recurring/edit", "Contributions")
 object DigiPackEditProfilePage extends IdentityPage("/digitalpack/edit", "Digital Pack")
-object ConsentJourneyPage extends IdentityPage("/consent", "Consent")
+
+sealed abstract class ConsentJourneyPage(id: String, journey: String) extends IdentityPage(id, "Consent") {
+  val journeyParam: String = journey
```
Comment (Calum-Campbell) [inline identity/app/controllers/EditProfileController.scala]: @jfsoul not the most elegant way to have access to the param, but does work..

Comment (TBonnin) [reply identity/app/controllers/EditProfileController.scala]: `..., val journey: String)` should do the trick :)
